src/github.com/getlantern/detour/detour.go

```diff
@@ -0,0 +1,353 @@
++-------------------------+-----------+-------------+-------------+-------------+-------------+
+(buf) = resend buffer
+tl = temporary whitelist
+wl = permanent whitelist
+* Operation will time out in TimeoutToDetour in initial state,
+and at system default or caller supplied deadline for other states;
+but in detoured state, it is considered as timeout if an operation takes longer than TimeoutToDetour.
+** DNS hijack is only checked at dial time.
+*** Connection is always detoured if the site is in tl or wl.
+*/
+package detour
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/getlantern/golog"
+)
+
+// if dial or read exceeded this timeout, we consider switch to detour
+// The value depends on OS and browser and defaults to 3s
+// For Windows XP, find TcpMaxConnectRetransmissions in
+// http://support2.microsoft.com/default.aspx?scid=kb;en-us;314053
+var TimeoutToDetour = 3 * time.Second
+
+var (
+	log = golog.LoggerFor("detour")
+
+	// instance of Detector
+	blockDetector atomic.Value
+)
+
+func init() {
+	blockDetector.Store(detectorByCountry(""))
+}
+
+type dialFunc func(network, addr string) (net.Conn, error)
+
+type Conn struct {
+	muConn sync.RWMutex
+	// the actual connection, will change so protect it
+	// can't user atomic.Value as the concrete type may vary
+	conn net.Conn
+
+	// don't access directly, use inState() and setState() instead
+	state uint32
+
+	// the function to dial detour if the site fails to connect directly
+	dialDetour dialFunc
+
+	// keep track of the total bytes read in this connection
+	readBytes int64
+
+	muLocalBuffer sync.Mutex
+	// localBuffer keep track of bytes sent through direct connection
+	// in initial state so we can resend them when detour
+	localBuffer bytes.Buffer
+
+	network, addr string
+	readDeadline  time.Time
+	writeDeadline time.Time
+}
+
+const (
+	stateInitial = iota
+	stateDirect
+	stateDetour
+)
+
+var statesDesc = []string{
+	"initially",
+	"directly",
+	"detoured",
+}
+
+// SetCountry sets the ISO 3166-1 alpha-2 country code to load country specific detection rules
+func SetCountry(country string) {
+	blockDetector.Store(detectorByCountry(country))
+}
+
+// Dialer returns a function with same signature of net.Dialer.Dial().
+func Dialer(d dialFunc) dialFunc {
+	return func(network, addr string) (conn net.Conn, err error) {
+		dc := &Conn{dialDetour: d, network: network, addr: addr}
+		if !whitelisted(addr) {
+			detector := blockDetector.Load().(*Detector)
+			dc.setState(stateInitial)
+			// always try direct connection first
+			dc.conn, err = net.DialTimeout(network, addr, TimeoutToDetour)
+			if err == nil {
+				if !detector.CheckConn(dc.conn) {
+					log.Tracef("Dial %s to %s succeeded", dc.stateDesc(), addr)
+					return dc, nil
+				}
+				log.Debugf("Dial %s to %s, dns hijacked, try detour", dc.stateDesc(), addr)
+			} else if detector.CheckError(err) {
```

Comment:
May want to use more meaningful name than CheckError, for example TamperingSuspected.